libs/log/logger.go

```diff
@@ -8,6 +8,7 @@ import (
 
 // Logger is what any Tendermint library should take.
 type Logger interface {
+	IsDebugging() bool
```

Comment:
Is there not perhaps another way to go about implementing this that doesn't involve having to check a flag from outside of the logger interface? e.g. some kind of lazy Sprintf wrapper:

```go
type lazySprintf struct {
    format string
    args   []interface{}
}

func LazySprintf(format string, args ...interface{}) *lazySprintf {
    return &lazySprintf { format, args }
}

func (l *lazySprintf) String() string {
    return fmt.Sprintf(l.format, l.args...)
}

// ...
logger.Debug(LazySprintf("some value %d", i))
```

Reply:
> Is there not perhaps another way to go about implementing this that doesn't involve having to check a flag from outside of the logger interface? e.g. some kind of lazy Sprintf wrapper:

I do like this idea though :)

Reply:
I definitely have a preference for an approach that doesn't require an outside flag or additional call. It's easier to maintain and guaranteed that people will get right in the future. The alternative is that we have to maintain a linter (or a human) that chastises a programmer for missing the 'IsDebugging' statement. Might as well just bake the conditional check directly into the [`Debug` call from tm logger](https://github.com/tendermint/tendermint/pull/8845/files#diff-5fc3b1a441999dc93e30dee6d9323528a40f96a563a09a777300bbd4dd217a40L63) make it easier to get right.

Reply:
LazySprintf and LazyBlockHash are in :)